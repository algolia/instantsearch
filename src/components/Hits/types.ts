import { SearchResults } from 'algoliasearch-helper';
import { BindEventForHits, SendEventForHits } from '../../lib/utils';
import { Hits } from '../../types';
import { HitsCSSClasses, HitsTemplates } from '../../widgets/hits/hits';

export type HitsProps = {
  cssClasses: HitsCSSClasses;
  hits: Hits;
  results: SearchResults;
  sendEvent: SendEventForHits;
  bindEvent: BindEventForHits;
  templateProps: {
    [key: string]: any;
    templates: HitsTemplates;
  };
};
