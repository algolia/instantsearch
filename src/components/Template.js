import React from 'react';

import curry from 'lodash/function/curry';
import cloneDeep from 'lodash/lang/cloneDeep';
import mapValues from 'lodash/object/mapValues';

import hogan from 'hogan.js';

import isEqual from 'lodash/lang/isEqual';

class Template extends React.Component {
  shouldComponentUpdate(nextProps) {
    return !isEqual(this.props.data, nextProps.data) || this.props.templateKey !== nextProps.templateKey;
  }

  render() {
    const useCustomCompileOptions = this.props.useCustomCompileOptions[this.props.templateKey];
    const compileOptions = useCustomCompileOptions ? this.props.templatesConfig.compileOptions : {};

    const content = renderTemplate({
      templates: this.props.templates,
      templateKey: this.props.templateKey,
      compileOptions,
      helpers: this.props.templatesConfig.helpers,
      data: transformData(this.props.transformData, this.props.templateKey, this.props.data)
    });

    if (content === null) {
      // Adds a noscript to the DOM but virtual DOM is null
      // See http://facebook.github.io/react/docs/component-specs.html#render
      return null;
    }

    if (React.isValidElement(content)) {
      return <div {...this.props.rootProps}>{content}</div>;
    }

    return <div {...this.props.rootProps} dangerouslySetInnerHTML={{__html: content}} />;
  }
}

Template.propTypes = {
  data: React.PropTypes.object,
  rootProps: React.PropTypes.object,
  templateKey: React.PropTypes.string,
  templates: React.PropTypes.objectOf(React.PropTypes.oneOfType([
    React.PropTypes.string,
    React.PropTypes.func
  ])),
  templatesConfig: React.PropTypes.shape({
    helpers: React.PropTypes.objectOf(React.PropTypes.func),
    // https://github.com/twitter/hogan.js/#compilation-options
    compileOptions: React.PropTypes.shape({
      asString: React.PropTypes.bool,
      sectionTags: React.PropTypes.arrayOf(React.PropTypes.shape({
        o: React.PropTypes.string,
        c: React.PropTypes.string
      })),
      delimiters: React.PropTypes.string,
      disableLambda: React.PropTypes.bool
    })
  }),
  transformData: React.PropTypes.oneOfType([
    React.PropTypes.func,
    React.PropTypes.objectOf(React.PropTypes.func)
  ]),
  useCustomCompileOptions: React.PropTypes.objectOf(React.PropTypes.bool)
};

Template.defaultProps = {
  data: {},
  useCustomCompileOptions: {},
  templates: {},
  templatesConfig: {}
};

function transformData(fn, templateKey, originalData) {
  if (!fn) {
    return originalData;
  }

  const clonedData = cloneDeep(originalData);

  let data;
  const typeFn = typeof fn;
  if (typeFn === 'function') {
    data = fn(clonedData);
  } else if (typeFn === 'object') {
    // ex: transformData: {hit, empty}
    if (fn[templateKey]) {
      data = fn[templateKey](clonedData);
    } else {
      // if the templateKey doesn't exist, just use the
      // original data
      data = originalData;
    }
  } else {
    throw new Error(`transformData must be a function or an object, was ${typeFn} (key : ${templateKey})`);
  }

  const dataType = typeof data;
  const expectedType = typeof originalData;
  if (dataType !== expectedType) {
    throw new Error(`\`transformData\` must return a \`${expectedType}\`, got \`${dataType}\`.`);
  }
  return data;
}

function renderTemplate({templates, templateKey, compileOptions, helpers, data}) {
  const template = templates[templateKey];
  const templateType = typeof template;
  const isTemplateString = templateType === 'string';
  const isTemplateFunction = templateType === 'function';

  if (!isTemplateString && !isTemplateFunction) {
    throw new Error(`Template must be 'string' or 'function', was '${templateType}' (key: ${templateKey})`);
  } else if (isTemplateFunction) {
    return template(data);
  } else {
    const transformedHelpers = transformHelpersToHogan(helpers, compileOptions, data);
    const preparedData = {...data, helpers: transformedHelpers};
    return hogan.compile(template, compileOptions).render(preparedData);
  }
}

// We add all our template helper methods to the template as lambdas. Note
// that lambdas in Mustache are supposed to accept a second argument of
// `render` to get the rendered value, not the literal `{{value}}`. But
// this is currently broken (see
// https://github.com/twitter/hogan.js/issues/222).
function transformHelpersToHogan(helpers, compileOptions, data) {
  return mapValues(helpers, method =>
    curry(function(text) {
      const render = value => hogan.compile(value, compileOptions).render(this);
      return method.call(data, text, render);
    })
  );
}

export default Template;
