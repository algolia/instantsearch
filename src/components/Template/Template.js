/** @jsx h */

import { h, Component } from 'preact';
import PropTypes from 'prop-types';
import { renderTemplate, isEqual } from '../../lib/utils';

class Template extends Component {
  shouldComponentUpdate(nextProps) {
    return (
      !isEqual(this.props.data, nextProps.data) ||
      this.props.templateKey !== nextProps.templateKey ||
      !isEqual(this.props.rootProps, nextProps.rootProps)
    );
  }

  render() {
    const RootTagName = this.props.rootTagName;
    const useCustomCompileOptions = this.props.useCustomCompileOptions[
      this.props.templateKey
    ];
    const compileOptions = useCustomCompileOptions
      ? this.props.templatesConfig.compileOptions
      : {};

    const content = renderTemplate({
      templates: this.props.templates,
      templateKey: this.props.templateKey,
      compileOptions,
      helpers: this.props.templatesConfig.helpers,
      data: this.props.data,
    });

    if (content === null) {
      // Adds a noscript to the DOM but virtual DOM is null
      // See http://facebook.github.io/react/docs/component-specs.html#render
      return null;
    }

    return (
      <RootTagName
        {...this.props.rootProps}
        dangerouslySetInnerHTML={{ __html: content }}
      />
    );
  }
}

Template.propTypes = {
  data: PropTypes.object,
  rootProps: PropTypes.object,
  rootTagName: PropTypes.string,
  templateKey: PropTypes.string,
  templates: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.string, PropTypes.func])
  ),
  templatesConfig: PropTypes.shape({
    helpers: PropTypes.objectOf(PropTypes.func),
    // https://github.com/twitter/hogan.js/#compilation-options
    compileOptions: PropTypes.shape({
      asString: PropTypes.bool,
      sectionTags: PropTypes.arrayOf(
        PropTypes.shape({
          o: PropTypes.string,
          c: PropTypes.string,
        })
      ),
      delimiters: PropTypes.string,
      disableLambda: PropTypes.bool,
    }),
  }),
  useCustomCompileOptions: PropTypes.objectOf(PropTypes.bool),
};

Template.defaultProps = {
  data: {},
  rootTagName: 'div',
  useCustomCompileOptions: {},
  templates: {},
  templatesConfig: {},
};

export default Template;
