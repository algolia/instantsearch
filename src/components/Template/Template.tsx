/** @jsx h */

import type { JSX } from 'preact';
import { h, Component } from 'preact';
import type { BindEventForHits, SendEventForHits } from '../../lib/utils';
import { renderTemplate, isEqual } from '../../lib/utils';
import type { PreparedTemplateProps } from '../../lib/utils/prepareTemplateProps';
import type { Templates } from '../../types';

const defaultProps = {
  data: {},
  rootTagName: 'div',
  useCustomCompileOptions: {},
  templates: {},
  templatesConfig: {},
};

type TemplateProps = {
  data?: Record<string, any>;
  rootProps?: Record<string, any>;
  rootTagName?: keyof JSX.IntrinsicElements;
  templateKey: string;
  bindEvent?: BindEventForHits;
  sendEvent?: SendEventForHits;
} & PreparedTemplateProps<Templates> &
  Readonly<typeof defaultProps>;

// @TODO: Template should be a generic and receive TData to pass to Templates (to avoid TTemplateData to be set as `any`)
class Template extends Component<TemplateProps> {
  public static readonly defaultProps = defaultProps;

  public shouldComponentUpdate(nextProps: TemplateProps) {
    return (
      !isEqual(this.props.data, nextProps.data) ||
      this.props.templateKey !== nextProps.templateKey ||
      !isEqual(this.props.rootProps, nextProps.rootProps)
    );
  }

  public render() {
    const RootTagName = this.props.rootTagName;

    const useCustomCompileOptions =
      this.props.useCustomCompileOptions[this.props.templateKey];
    const compileOptions = useCustomCompileOptions
      ? this.props.templatesConfig.compileOptions
      : {};

    const content = renderTemplate({
      templates: this.props.templates,
      templateKey: this.props.templateKey,
      compileOptions,
      helpers: this.props.templatesConfig.helpers,
      data: this.props.data,
      bindEvent: this.props.bindEvent,
      sendEvent: this.props.sendEvent,
    });

    if (content === null) {
      // Adds a noscript to the DOM but virtual DOM is null
      // See http://facebook.github.io/react/docs/component-specs.html#render
      return null;
    }

    if (typeof content === 'object') {
      return <RootTagName {...this.props.rootProps}>{content}</RootTagName>;
    }

    return (
      <RootTagName
        {...this.props.rootProps}
        dangerouslySetInnerHTML={{ __html: content }}
      />
    );
  }
}

export default Template;
