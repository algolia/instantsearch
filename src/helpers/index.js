export { default as highlight } from './highlight.js';
export { default as snippet } from './snippet.js';
