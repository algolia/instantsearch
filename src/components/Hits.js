import PropTypes from 'prop-types';
import React, { Component } from 'preact-compat';
import map from 'lodash/map';
import Template from './Template';
import hasKey from 'lodash/has';
import cx from 'classnames';

class Hits extends Component {
  renderWithResults() {
    const renderedHits = map(this.props.hits, (hit, position) => {
      const data = {
        ...hit,
        __hitIndex: position,
      };
      return (
        <Template
          data={data}
          key={data.objectID}
          rootProps={{ className: this.props.cssClasses.item }}
          templateKey="item"
          {...this.props.templateProps}
        />
      );
    });

    return <div className={this.props.cssClasses.root}>{renderedHits}</div>;
  }

  renderAllResults() {
    const className = cx(
      this.props.cssClasses.root,
      this.props.cssClasses.allItems
    );

    return (
      <Template
        data={this.props.results}
        rootProps={{ className }}
        templateKey="allItems"
        {...this.props.templateProps}
      />
    );
  }

  renderNoResults() {
    const className = cx(
      this.props.cssClasses.root,
      this.props.cssClasses.empty
    );
    return (
      <Template
        data={this.props.results}
        rootProps={{ className }}
        templateKey="empty"
        {...this.props.templateProps}
      />
    );
  }

  render() {
    const hasResults = this.props.results.hits.length > 0;
    const hasAllItemsTemplate = hasKey(
      this.props,
      'templateProps.templates.allItems'
    );

    if (!hasResults) {
      return this.renderNoResults();
    }

    // If a allItems template is defined, it takes precedence over our looping
    // through hits
    if (hasAllItemsTemplate) {
      return this.renderAllResults();
    }

    return this.renderWithResults();
  }
}

Hits.propTypes = {
  cssClasses: PropTypes.shape({
    root: PropTypes.string,
    item: PropTypes.string,
    allItems: PropTypes.string,
    empty: PropTypes.string,
  }),
  hits: PropTypes.array,
  results: PropTypes.object,
  templateProps: PropTypes.object.isRequired,
};

Hits.defaultProps = {
  results: { hits: [] },
};

export default Hits;
