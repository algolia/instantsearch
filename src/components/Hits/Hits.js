/** @jsx h */

import { h } from 'preact';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Template from '../Template/Template';

const Hits = ({ results, hits, cssClasses, templateProps }) => {
  if (results.hits.length === 0) {
    return (
      <Template
        {...templateProps}
        templateKey="empty"
        rootProps={{
          className: cx(cssClasses.root, cssClasses.emptyRoot),
        }}
        data={results}
      />
    );
  }

  return (
    <div className={cssClasses.root}>
      <ol className={cssClasses.list}>
        {hits.map((hit, position) => (
          <Template
            {...templateProps}
            templateKey="item"
            rootTagName="li"
            rootProps={{ className: cssClasses.item }}
            key={hit.objectID}
            data={{
              ...hit,
              __hitIndex: position,
            }}
          />
        ))}
      </ol>
    </div>
  );
};

Hits.propTypes = {
  cssClasses: PropTypes.shape({
    root: PropTypes.string.isRequired,
    emptyRoot: PropTypes.string.isRequired,
    item: PropTypes.string.isRequired,
    list: PropTypes.string.isRequired,
  }).isRequired,
  hits: PropTypes.array.isRequired,
  results: PropTypes.object.isRequired,
  templateProps: PropTypes.object.isRequired,
};

Hits.defaultProps = {
  results: { hits: [] },
  hits: [],
};

export default Hits;
