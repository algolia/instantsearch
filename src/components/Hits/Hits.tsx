/** @jsx h */

import { h } from 'preact';
import cx from 'classnames';
import Template from '../Template/Template';
import type { SearchResults } from 'algoliasearch-helper';
import type {
  BindEventForHits,
  SendEventForHits,
  PreparedTemplateProps,
} from '../../lib/utils';
import type { ComponentCSSClasses, Hit } from '../../types';
import type { HitsCSSClasses, HitsTemplates } from '../../widgets/hits/hits';

export type HitsComponentCSSClasses = ComponentCSSClasses<HitsCSSClasses>;
export type HitsComponentTemplates = Required<HitsTemplates>;

export type HitsProps = {
  results: SearchResults;
  hits: Hit[];
  sendEvent?: SendEventForHits;
  bindEvent?: BindEventForHits;
  cssClasses: HitsComponentCSSClasses;
  templateProps: PreparedTemplateProps<HitsComponentTemplates>;
};

const Hits = ({
  results,
  hits,
  bindEvent,
  sendEvent,
  cssClasses,
  templateProps,
}: HitsProps) => {
  if (results.hits.length === 0) {
    return (
      <Template
        {...templateProps}
        templateKey="empty"
        rootProps={{
          className: cx(cssClasses.root, cssClasses.emptyRoot),
        }}
        data={results}
      />
    );
  }

  return (
    <div className={cssClasses.root}>
      <ol className={cssClasses.list}>
        {hits.map((hit, index) => (
          <Template
            {...templateProps}
            templateKey="item"
            rootTagName="li"
            rootProps={{ className: cssClasses.item }}
            key={hit.objectID}
            data={{
              ...hit,
              __hitIndex: index,
            }}
            bindEvent={bindEvent}
            sendEvent={sendEvent}
          />
        ))}
      </ol>
    </div>
  );
};

export default Hits;
