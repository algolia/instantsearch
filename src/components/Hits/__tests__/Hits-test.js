/** @jsx h */

import { h } from 'preact';
import { shallow, mount } from 'enzyme';
import { highlight } from '../../../helpers';
import { TAG_REPLACEMENT } from '../../../lib/escape-highlight';
import Template from '../../Template/Template';
import Hits from '../Hits';

describe('Hits', () => {
  const cssClasses = {
    root: 'root',
    emptyRoot: 'emptyRoot',
    item: 'item',
    list: 'list',
  };

  function shallowRender(extraProps = {}) {
    const props = {
      cssClasses,
      templateProps: {},
      ...extraProps,
    };

    return shallow(<Hits {...props} />);
  }

  describe('no results', () => {
    it('should use the empty template if no results', () => {
      const props = {
        results: {
          hits: [],
        },
        hits: [],
        cssClasses,
      };

      const wrapper = shallowRender(props);

      expect(wrapper.props().templateKey).toEqual('empty');
    });

    it('should set the empty CSS class when no results', () => {
      const props = {
        results: {
          hits: [],
        },
        cssClasses,
      };

      const wrapper = shallowRender(props);

      expect(wrapper.props().rootProps.className).toContain('root');
    });
  });

  describe('individual item templates', () => {
    it('should add an item template for each hit', () => {
      const hits = [
        {
          objectID: 'one',
          foo: 'bar',
        },
        {
          objectID: 'two',
          foo: 'baz',
        },
      ];
      const props = {
        results: { hits },
        hits,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find(Template);

      expect(wrapper).toHaveLength(2);
      expect(wrapper.at(0).props().templateKey).toEqual('item');
    });

    it('should set the item class to each item', () => {
      const hits = [
        {
          objectID: 'one',
          foo: 'bar',
        },
      ];
      const props = {
        results: { hits },
        hits,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find(Template);

      expect(wrapper.props().rootProps.className).toContain('item');
    });

    it('should wrap the items in a root div element', () => {
      const props = {
        results: {
          hits: [
            {
              objectID: 'one',
              foo: 'bar',
            },
            {
              objectID: 'two',
              foo: 'baz',
            },
          ],
        },
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props);

      expect(wrapper.name()).toEqual('div');
      expect(wrapper.props().className).toContain('root');
    });

    it('should pass each result data to each item template', () => {
      const hits = [
        {
          objectID: 'one',
          foo: 'bar',
        },
        {
          objectID: 'two',
          foo: 'baz',
        },
      ];
      const props = {
        results: { hits },
        hits,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find({ templateKey: 'item' });

      expect(wrapper.at(0).props().data.foo).toEqual('bar');
      expect(wrapper.at(1).props().data.foo).toEqual('baz');
    });

    it('should add the __hitIndex in the list to each item', () => {
      const hits = [
        {
          objectID: 'one',
          foo: 'bar',
        },
        {
          objectID: 'two',
          foo: 'baz',
        },
      ];
      const props = {
        results: { hits },
        hits,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find({ templateKey: 'item' });

      expect(wrapper.at(0).props().data.__hitIndex).toEqual(0);
      expect(wrapper.at(1).props().data.__hitIndex).toEqual(1);
    });

    it('should use the objectID as the DOM key', () => {
      const hits = [
        {
          objectID: 'BAR',
          foo: 'bar',
        },
        {
          objectID: 'BAZ',
          foo: 'baz',
        },
      ];
      const props = {
        results: { hits },
        hits,
        templateProps: {
          templates: {
            item: 'one item',
          },
        },
        cssClasses,
      };

      const wrapper = shallowRender(props).find({ templateKey: 'item' });

      expect(wrapper.at(0).key()).toEqual('BAR');
      expect(wrapper.at(1).key()).toEqual('BAZ');
    });
  });

  describe('markup', () => {
    it('should render <Hits />', () => {
      const hits = [
        {
          objectID: 'one',
          foo: 'bar',
        },
        {
          objectID: 'two',
          foo: 'baz',
        },
      ];

      const props = {
        results: { hits },
        hits,
        templateProps: {
          templates: {
            item: 'item',
          },
        },
        cssClasses,
      };

      const wrapper = mount(<Hits {...props} />);

      expect(wrapper).toMatchSnapshot();
    });

    it('should render <Hits /> without highlight function', () => {
      const hits = [
        {
          objectID: 'one',
          name: 'name 1',
          _highlightResult: {
            name: {
              value: `${TAG_REPLACEMENT.highlightPreTag}name 1${TAG_REPLACEMENT.highlightPostTag}`,
            },
          },
        },
        {
          objectID: 'two',
          name: 'name 2',
          _highlightResult: {
            name: {
              value: `${TAG_REPLACEMENT.highlightPreTag}name 2${TAG_REPLACEMENT.highlightPostTag}`,
            },
          },
        },
      ];

      const props = {
        results: { hits },
        hits,
        templateProps: {
          templates: {
            item(hit) {
              return highlight({
                attribute: 'name',
                hit,
              });
            },
          },
        },
        cssClasses,
      };

      const wrapper = mount(<Hits {...props} />);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
