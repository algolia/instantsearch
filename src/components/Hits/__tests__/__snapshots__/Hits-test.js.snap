// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hits markup should render <Hits /> 1`] = `
<div
  className="root"
>
  <ol
    className="list"
  >
    <li
      className="item"
      dangerouslySetInnerHTML={
        Object {
          "__html": "item",
        }
      }
    />
    <li
      className="item"
      dangerouslySetInnerHTML={
        Object {
          "__html": "item",
        }
      }
    />
  </ol>
</div>
`;

exports[`Hits markup should render <Hits /> without highlight function 1`] = `
<div
  className="root"
>
  <ol
    className="list"
  >
    <li
      className="item"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<mark class=\\"ais-HighLight-highlighted\\">name 1</mark>",
        }
      }
    />
    <li
      className="item"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<mark class=\\"ais-HighLight-highlighted\\">name 2</mark>",
        }
      }
    />
  </ol>
</div>
`;
